packages/core/src/types/wrapperObject.ts

```diff
@@ -0,0 +1,28 @@
+import { InterpreterContext } from "../runtime/interpreter/interpreterContext.js";
+import { WrapperObject } from "../runtime/objects/wrapperObject.js";
+import { Type } from "./base.js";
+
+/**
+ * Generates a wrapper object type
+ *
+ * @param name the name of the type
+ * @param matches validates the wrapped object
+ * @returns the generated type
+ */
+export function wrapperObjectType(
+    name: string,
+    matches: (value: WrapperObject<any>, context: InterpreterContext) => boolean
+): Type {
+    return {
+        name: () => name,
+        matches(value, context) {
+            if (!(value instanceof WrapperObject) || !matches(value, context)) {
+                return {
+                    expected: this,
+                    path: []
+                };
+            }
+            return true;
```

Comment:
Shouldn't this function return a boolean?
Why doesn't it return a boolean in the error case?

Reply:
```ts
/**
 * The result of the match function, either true or an object with the current path and reson why it did not match
 */
export type MatchesResult =
    | true
    | {
          path: string[];
          expected: Type;
      };
```